### PR TITLE
Migrate TermsQuery to TermInSetQuery

### LINKED ISSUE #23201: Remove usage of `TermsQuery`
`TermsQuery` is now deprecated in Lucene and has been replaced by `TermInSetQuery`, which is slightly different as it operates on a single field. We should migrate our usage of `TermsQuery` to `TermInSetQuery`.

### PR DESCRIPTION
**Pull request by martinscholz83**
Fix #23201
This PR replace the usage of `TermsQuery` with `TermInSetQuery`. There are two tests i have commented out because i think they will no longer be needed.

### CODE REVIEW

**Comment by rjernst** [comment]
@elasticmachine test this please

File: core/src/test/java/org/elasticsearch/index/mapper/KeywordFieldTypeTests.java [hunk 1]
```diff
@@ -106,18 +106,18 @@ protected TokenStream normalize(String fieldName, TokenStream in) {
...
-        assertEquals(new TermsQuery(new Term("field", "foo"), new Term("field", "bar")),
+        assertEquals(new TermInSetQuery(new Term("field", "foo"), new Term("field", "bar")),
                 ft.termsQuery(Arrays.asList("foo", "bar"), null));
 
         ft.setIndexOptions(IndexOptions.NONE);
         IllegalArgumentException e = expectThrows(IllegalArgumentException.class,
                 () -> ft.termsQuery(Arrays.asList("foo", "bar"), null));
         assertEquals("Cannot search on field [field] since it is not indexed.", e.getMessage());
-    }
+    }*/
```
Comment (jpountz) [inline core/src/test/java/org/elasticsearch/index/mapper/KeywordFieldTypeTests.java]: why did you comment out this test?

File: core/src/test/java/org/elasticsearch/index/mapper/TextFieldTypeTests.java [hunk 2]
```diff
@@ -82,18 +82,19 @@ public void testTermQuery() {
         assertEquals("Cannot search on field [field] since it is not indexed.", e.getMessage());
     }
 
-    public void testTermsQuery() {
+    //We still need this test??
```
Comment (jpountz) [inline core/src/test/java/org/elasticsearch/index/mapper/TextFieldTypeTests.java]: yes, please. Lucene renamed the `TermsQuery` to `TermInSetQuery` but we plan on keeping support for the `terms` query in Elasticsearch, which is what this test is testing.

File: core/src/test/java/org/elasticsearch/index/query/HasChildQueryBuilderTests.java [hunk 3]
```diff
@@ -240,18 +240,14 @@ static void assertLateParsingQuery(Query query, String type, String id) throws I
         assertThat(booleanQuery.clauses().size(), equalTo(2));
         //check the inner ids query, we have to call rewrite to get to check the type it's executed against
         assertThat(booleanQuery.clauses().get(0).getOccur(), equalTo(BooleanClause.Occur.MUST));
-        assertThat(booleanQuery.clauses().get(0).getQuery(), instanceOf(TermsQuery.class));
-        TermsQuery termsQuery = (TermsQuery) booleanQuery.clauses().get(0).getQuery();
-        // we need to rewrite once for TermsQuery -> TermInSetQuery and than againt TermInSetQuery -> ConstantScoreQuery
+        assertThat(booleanQuery.clauses().get(0).getQuery(), instanceOf(TermInSetQuery.class));
+        TermInSetQuery termsQuery = (TermInSetQuery) booleanQuery.clauses().get(0).getQuery();
+        // we need to rewrite once for TermInSetQuery -> TermInSetQuery and than againt TermInSetQuery -> ConstantScoreQuery
```
Comment (jpountz) [inline core/src/test/java/org/elasticsearch/index/query/HasChildQueryBuilderTests.java]: this comment is outdated now, can you remove one of the rewrites on the line below?

**Review by jpountz — CHANGES_REQUESTED** [review]
I left some comments.

**Comment by jpountz** [comment]
@elasticmachine test this please

**Comment by maddin2016** [comment]
Hi @jpountz, i have rewrite the tests. I was litte bit confused by the tests name. I thought these tests where about the constructor of `TermsQuery`. So now all should be fine.

**Review by jpountz — APPROVED** [review]
Thanks, the diff looks great now.

**Comment by jpountz** [comment]
@elasticmachine test this please

**Comment by maddin2016** [comment]
👍

**Comment by jpountz** [comment]
Thanks @maddin2016 !

**Pull request merged**